flybywiresim/a32nx
# feat(efb): fix landing calculator issues and allow for unit toggle

Fixes #5734 and #5723

## Summary of Changes

- General: Added a typescript unit conversion class similar to NXUnit that is synced to the `CONFIG_USING_METRIC_UNIT` setting.
- Landing calculator: Added unit toggles for "temperature", "pressure", "runway LDA" and "weight" inputs. The initial values and subsequent conversions make use of the new Units class.
- Landing calculator: Fixed the "Clear" button.
- Landing calculator: Fixed the "Wind Magnitude" input.
- Landing calculator: Made a separate button (from current *"Get METAR"*) to sync the weight.
- Landing calculator: Synced the displayed landing distance unit to the new "runway LDA" unit toggle *(m | ft)*. Also added a space between values and units, in the interests of good taste.
- Landing calculator: Made all inputs initially empty (by having *undefined* values), so that the placeholders can be seen.

## Screenshots (if necessary)
Initial landing calculator page (using metric system):
![image](https://user-images.githubusercontent.com/30361843/138364587-c498120a-9bb9-49f7-8873-2fc586938e3c.png)
Filling in weight + runway LDA in ft:
![image](https://user-images.githubusercontent.com/30361843/138364595-862401b8-c16a-4595-977d-a66e7e898eae.png)
Calculated values, length in m:
![image](https://user-images.githubusercontent.com/30361843/138364655-08fe7f6a-446c-4d5f-8164-9d0e756045b8.png)

## References

## Additional context

Discord username (if different from GitHub):

- @ghitier -> discord: `bk67`

## Testing instructions

Check that the landing performance page initially uses the units you select in Settings -> Aircraft -> Weight Unit (`kg` corresponds to metric units), when you change that setting the toggles should reflect that choice.

> Except for the Temperature which should initially always be in Celsius (it's the unit METARs use)

If you're able to use the Coherent debugger (if not don't worry), try to change the unit setting while looking at the EFB page:

- Open the debugger
- Choose the CDU
- In the console, run:
  - `NXDataStore.set('CONFIG_USING_METRIC_UNIT', '1')` to choose metric,
  - or `NXDataStore.set('CONFIG_USING_METRIC_UNIT', '0')` to choose imperial.
- Values should update in real-time on the EFB.

---

When you go to the landing calculator all inputs should initially be empty. The *"Calculate"* button will initially be greyed out, it should become enabled only once all fields are filled.



If you use the *"Get METAR"* button, make sure the weight doesn't get filled in anymore.

Check that the when you toggle between units, the value is converted accurately.

Use the button (down arrow icon) right of the weight label to auto-fill the input. Make the value is coherent with your currents aircraft weight. 

Make sure to play around with the different values to see if the landing distance gets updated coherently.

Use the *"Clear"* button and check all inputs get back to an empty state (placeholders visible).



## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
2. Click on the **Checks** tab on the PR
3. On the left side, click on the bottom **PR** tab
4. Click on the **A32NX** download link at the bottom of the page
